src/grids/linear_grid.jl

```diff
@@ -1,217 +1,218 @@
 
 """
 Internal function that works on padded arrays
 """
 function _convolution!(grid::LinearGrid, out::AbstractVector,
-    u::AbstractVector,v::AbstractVector)
-  (; padded, ft_work1, ft_work2, ft, ft⁻¹, conv_shift) = grid.op
-  
-  assert_applicable(ft⁻¹.p, ft_work2, out)
-  unsafe_execute!(ft, _zeropad!(padded, u), ft_work1)
-  unsafe_execute!(ft, _zeropad!(padded, v), ft_work2)
-  @. ft_work2 *= ft_work1 * conv_shift
-  unsafe_execute!(ft⁻¹.p, ft_work2, out)
-  
-  return out
+    u::AbstractVector, v::AbstractVector)
+    (; padded, ft_work1, ft_work2, ft_p, bft_p, ft_arr) = grid.op
+
+    assert_applicable(bft_p, ft_work2, out)
+    unsafe_execute!(ft_p, _zeropad!(padded, u), ft_work1)
+    unsafe_execute!(ft_p, _zeropad!(padded, v), ft_work2)
+    @. ft_work2 *= ft_work1 * ft_arr
+
+    unsafe_execute!(bft_p, ft_work2, out)
+
+    out .*= grid.d / length(padded)
+
+    return out
 end
 
 """
 Convolution - takes non-padded arrays as input (and output)
 """
 convolution!(grid::LinearGrid, out::AbstractVector,
-  u::AbstractVector,v::AbstractVector
-  ) = out .= @views _convolution!(grid, grid.op.ft⁻¹_work, u, v)[1:grid.len]
+    u::AbstractVector, v::AbstractVector
+) = out .= @views _convolution!(grid, grid.op.bft_work, u, v)[grid.op.data_range]
 
 convolution(grid::LinearGrid, u::AbstractVector, v::AbstractVector
-  ) = convolution!(grid, zero(u), u, v)
+) = convolution!(grid, zeros(eltype(grid.op.padded), length(u)), u, v)
 
 """
 Internal function that works on padded arrays
 """
 function _crosscorrelation!(grid::LinearGrid, out::AbstractVector,
     u::AbstractVector, v::AbstractVector)
-  (; padded, ft_work1, ft_work2, ft, ft⁻¹, corr_shift) = grid.op
-  
-  assert_applicable(ft⁻¹.p, ft_work2, out)
-  unsafe_execute!(ft, _zeropad!(padded, u), ft_work1)
-  unsafe_execute!(ft, _zeropad!(padded, v), ft_work2)
-  @. ft_work2 *= conj(ft_work1 * corr_shift)
-  unsafe_execute!(ft⁻¹.p, ft_work2, out)
-  
-  return out
+    (; padded, ft_work1, ft_work2, ft_p, bft_p, ft_arr) = grid.op
+
+    assert_applicable(bft_p, ft_work2, out)
+    unsafe_execute!(ft_p, _zeropad!(padded, u), ft_work1)
+    unsafe_execute!(ft_p, _zeropad!(padded, v), ft_work2)
+
+    @. ft_work2 *= conj(ft_work1) * ft_arr
+    unsafe_execute!(bft_p, ft_work2, out)
+
+
+    out .*= grid.d / length(padded)
+
+    return out
 end
 
 """
 Cross-correlation - takes non-padded arrays as input (and output)
 """
 crosscorrelation!(grid::LinearGrid, out::AbstractVector,
-  u::AbstractVector, v::AbstractVector
-  ) = out .= @views _crosscorrelation!(grid, grid.op.ft⁻¹_work, u, v)[1:grid.len]
+    u::AbstractVector, v::AbstractVector
+) = out .= @views _crosscorrelation!(grid, grid.op.bft_work, u, v)[grid.op.data_range]
 
 crosscorrelation(grid::LinearGrid, u::AbstractVector, v::AbstractVector
-  ) = crosscorrelation!(grid, zero(u), u, v)
+) = crosscorrelation!(grid, zeros(eltype(grid.op.padded), length(u)), u, v)
```

Comment:
We should make both u and v have the same types AbstractVector{T}. But The output should have the same data structure and size as the input, no? This function is on non-padded arrays.

Reply:
yes, this needs to be done in more detail. If one inputs a real array u, due to complex_op = true by default, the output must be complex. This was just a quick fix..

Reply:
Right, we should check actually what happens when we give a real function to a complex_op=true grid in general and make this stable.